### ISSUE TITLE
build failed with ffmpeg version 6.1

### ISSUE THREAD

**Issue by trungtv94** [scriptituk/xfade-easing#9]
Hello, thank you for your amazing work. 
I successed to build with ffmpeg version 7.1 but my hardware can't use this version so I tried to build ffmpeg version 6.1 with configure:
./configure --enable-gpl --enable-libx264 --enable-libx265 --enable-libvpx --enable-libmp3lame --enable-opengl --enable-cuda --enable-cuvid --enable-nvenc

I got error when make:

libavfilter/xfade-easing.h:2209:20: error: implicit declaration of function ‘ff_filter_link’; did you mean ‘avfilter_link’? [-Werror=implicit-function-declaration]
 2209 |     AVRational r = ff_filter_link(l)->frame_rate;
      |                    ^~~~~~~~~~~~~~
      |                    avfilter_link
libavfilter/xfade-easing.h:2209:37: error: invalid type argument of ‘->’ (have ‘int’)
 2209 |     AVRational r = ff_filter_link(l)->frame_rate;
      |                                     ^~
libavfilter/xfade-easing.h:2221:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 2221 |     XTransition e = { .k = k, .ratio = (float)l->w / l->h, .is_rgb = s->is_rgb };
      |     ^~~~~~~~~~~
libavfilter/xfade-easing.h: In function ‘inverted_page_curl’:
libavfilter/xfade-easing.h:2283:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 2283 |     vec2 point = add2(rot(e->p, ang), o1), p;
      |     ^~~~
libavfilter/xfade-easing.h:2308:9: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 2308 |         vec2 q = add2(rot(p, -ang), o2);
      |         ^~~~
libavfilter/xfade-easing.h:2317:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 2317 |     float hitAngleMod = glmod(hitAngle, M_2PIf);
      |     ^~~~~
libavfilter/xfade-easing.h:2324:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 2324 |     float dx = (point.x < 0) ? -point.x : (point.x > 1) ? point.x - 1 : (point.x > P5f) ? 1 - point.x : point.x;
      |     ^~~~~
libavfilter/xfade-easing.h:2340:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 2340 |     float g = colour.p0;
      |     ^~~~~
libavfilter/xfade-easing.h: In function ‘solve_cubic_bezier’:
libavfilter/xfade-easing.h:2394:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 2394 |     const float cx = 3 * x1, bx = 3 * (x2 - x1) - cx, ax = 1 - cx - bx,
      |     ^~~~~
libavfilter/xfade-easing.h:2404:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 2404 |     float t0 = 0, t1 = 0;
      |     ^~~~~
libavfilter/xfade-easing.h:2414:5: warning: ISO C90 forbids mixed declarations and code [-Wdeclaration-after-statement]
 2414 |     const float kBezierEpsilon = 1e-7f;
      |     ^~~~~
libavfilter/vf_xfade.c: In function ‘config_output’:
libavfilter/vf_xfade.c:2054:5: error: unknown type name ‘FilterLink’; did you mean ‘AVFilterLink’?
 2054 |     FilterLink      *inl0 = ff_filter_link(inlink0);
      |     ^~~~~~~~~~
      |     AVFilterLink
libavfilter/vf_xfade.c:2054:29: warning: initialization of ‘int *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 2054 |     FilterLink      *inl0 = ff_filter_link(inlink0);
      |                             ^~~~~~~~~~~~~~
libavfilter/vf_xfade.c:2055:5: error: unknown type name ‘FilterLink’; did you mean ‘AVFilterLink’?
 2055 |     FilterLink      *inl1 = ff_filter_link(inlink1);
      |     ^~~~~~~~~~
      |     AVFilterLink
libavfilter/vf_xfade.c:2055:29: warning: initialization of ‘int *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 2055 |     FilterLink      *inl1 = ff_filter_link(inlink1);
      |                             ^~~~~~~~~~~~~~
libavfilter/vf_xfade.c:2056:5: error: unknown type name ‘FilterLink’; did you mean ‘AVFilterLink’?
 2056 |     FilterLink        *ol = ff_filter_link(outlink);
      |     ^~~~~~~~~~
      |     AVFilterLink
libavfilter/vf_xfade.c:2056:29: warning: initialization of ‘int *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 2056 |     FilterLink        *ol = ff_filter_link(outlink);
      |                             ^~~~~~~~~~~~~~
libavfilter/vf_xfade.c:2079:14: error: request for member ‘frame_rate’ in something not a structure or union
 2079 |     if (!inl0->frame_rate.num || !inl0->frame_rate.den) {

Can you help me to correct it?

**Comment by scriptituk** [comment]
Sorry about that.

Please change line 2341 of `xfade-easing.h` to
```
#if LIBAVFILTER_VERSION_INT < AV_VERSION_INT(10, 4, 100)
```
Thanks for raising this.

**Comment by scriptituk** [comment]
And to eliminate the `-Wdeclaration-after-statement` warnings compile using
```
make ECFLAGS=-Wno-declaration-after-statement
```

**Comment by trungtv94** [comment]
Thank you so much for your support, with changing follow your guide.
it passed error above. But still have 2nd error:

CC      libavfilter/vf_xmedian.o
CC      libavfilter/vf_yadif.o
In file included from libavfilter/vf_xfade.c:2013:
libavfilter/xfade-easing.h: In function ‘config_xfade_easing’:
libavfilter/xfade-easing.h:2209:20: error: implicit declaration of function ‘ff_filter_link’; did you mean ‘avfilter_link’? [-Werror=implicit-function-declaration]
 2209 |     AVRational r = ff_filter_link(l)->frame_rate;
      |                    ^~~~~~~~~~~~~~
      |                    avfilter_link
libavfilter/xfade-easing.h:2209:37: error: invalid type argument of ‘->’ (have ‘int’)
 2209 |     AVRational r = ff_filter_link(l)->frame_rate;
      |                                     ^~
libavfilter/xfade-easing.h: At top level:
libavfilter/xfade-easing.h:2341: error: unterminated #if
 2341 |     #if LIBAVFILTER_VERSION_INT < AV_VERSION_INT(10, 4, 100)
      |
libavfilter/vf_xfade.c: In function ‘config_output’:
libavfilter/vf_xfade.c:2054:5: error: unknown type name ‘FilterLink’; did you mean ‘AVFilterLink’?
 2054 |     FilterLink      *inl0 = ff_filter_link(inlink0);
      |     ^~~~~~~~~~
      |     AVFilterLink
libavfilter/vf_xfade.c:2054:29: warning: initialization of ‘int *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 2054 |     FilterLink      *inl0 = ff_filter_link(inlink0);
      |                             ^~~~~~~~~~~~~~
libavfilter/vf_xfade.c:2055:5: error: unknown type name ‘FilterLink’; did you mean ‘AVFilterLink’?
 2055 |     FilterLink      *inl1 = ff_filter_link(inlink1);
      |     ^~~~~~~~~~
      |     AVFilterLink
libavfilter/vf_xfade.c:2055:29: warning: initialization of ‘int *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 2055 |     FilterLink      *inl1 = ff_filter_link(inlink1);
      |                             ^~~~~~~~~~~~~~
libavfilter/vf_xfade.c:2056:5: error: unknown type name ‘FilterLink’; did you mean ‘AVFilterLink’?
 2056 |     FilterLink        *ol = ff_filter_link(outlink);
      |     ^~~~~~~~~~
      |     AVFilterLink
libavfilter/vf_xfade.c:2056:29: warning: initialization of ‘int *’ from ‘int’ makes pointer from integer without a cast [-Wint-conversion]
 2056 |     FilterLink        *ol = ff_filter_link(outlink);
      |                             ^~~~~~~~~~~~~~
libavfilter/vf_xfade.c:2079:14: error: request for member ‘frame_rate’ in something not a structure or union
 2079 |     if (!inl0->frame_rate.num || !inl0->frame_rate.den) {
      |              ^~
libavfilter/vf_xfade.c:2079:39: error: request for member ‘frame_rate’ in something not a structure or union
 2079 |     if (!inl0->frame_rate.num || !inl0->frame_rate.den) {
      |                                       ^~
libavfilter/vf_xfade.c:2081:58: error: request for member ‘frame_rate’ in something not a structure or union
 2081 |                "current rate of %d/%d is invalid\n", inl0->frame_rate.num, inl0->frame_rate.den);
      |                                                          ^~
libavfilter/vf_xfade.c:2081:80: error: request for member ‘frame_rate’ in something not a structure or union
 2081 |                "current rate of %d/%d is invalid\n", inl0->frame_rate.num, inl0->frame_rate.den);
      |                                                                                ^~
libavfilter/vf_xfade.c:2085:13: error: request for member ‘frame_rate’ in something not a structure or union
 2085 |     if (inl0->frame_rate.num != inl1->frame_rate.num ||
      |             ^~
libavfilter/vf_xfade.c:2085:37: error: request for member ‘frame_rate’ in something not a structure or union
 2085 |     if (inl0->frame_rate.num != inl1->frame_rate.num ||
      |                                     ^~
libavfilter/vf_xfade.c:2086:13: error: request for member ‘frame_rate’ in something not a structure or union
 2086 |         inl0->frame_rate.den != inl1->frame_rate.den) {
      |             ^~
libavfilter/vf_xfade.c:2086:37: error: request for member ‘frame_rate’ in something not a structure or union
 2086 |         inl0->frame_rate.den != inl1->frame_rate.den) {
      |                                     ^~
libavfilter/vf_xfade.c:2090:45: error: request for member ‘frame_rate’ in something not a structure or union
 2090 |                ctx->input_pads[0].name, inl0->frame_rate.num, inl0->frame_rate.den,
      |                                             ^~
libavfilter/vf_xfade.c:2090:67: error: request for member ‘frame_rate’ in something not a structure or union
 2090 |                ctx->input_pads[0].name, inl0->frame_rate.num, inl0->frame_rate.den,
      |                                                                   ^~
libavfilter/vf_xfade.c:2091:45: error: request for member ‘frame_rate’ in something not a structure or union
 2091 |                ctx->input_pads[1].name, inl1->frame_rate.num, inl1->frame_rate.den);
      |                                             ^~
libavfilter/vf_xfade.c:2091:67: error: request for member ‘frame_rate’ in something not a structure or union
 2091 |                ctx->input_pads[1].name, inl1->frame_rate.num, inl1->frame_rate.den);
      |                                                                   ^~
libavfilter/vf_xfade.c:2099:7: error: request for member ‘frame_rate’ in something not a structure or union
 2099 |     ol->frame_rate = inl0->frame_rate;
      |       ^~
libavfilter/vf_xfade.c:2099:26: error: request for member ‘frame_rate’ in something not a structure or union
 2099 |     ol->frame_rate = inl0->frame_rate;
      |                          ^~


Please help me to check this. Thank you.

**Comment by scriptituk** [comment]
Your line number `xfade-easing.h:2209` does not correspond with the repo file, it should be 2344, perhaps it's old.

Your `vf_xfade.c` errors indicate you are using v7.1 `vf_xfade.c` which is not compatible with v6.1.

I have updated [xfade-easing.h](../blob/main/src/xfade-easing.h) with the version fix.

If you download that file and [vf_xfade.c](../blob/main/src/v6.1/vf_xfade.c) to your 6.1 `libavfilter/` directory it should compile.

I have just built v6.1.2 without problem.

The version fix hides `ff_filter_link` from pre-v7.1 builds.
